lib/s3.py

```diff
@@ -111,14 +111,20 @@ def sign_curl(url: urllib.parse.ParseResult, method='GET', headers={}, checksum=
 
 def urlopen(url: urllib.parse.ParseResult, method='GET', headers={}, data=b'') -> IO:
     """Same as sign_request() but calls urlopen() on the result"""
-    headers = sign_request(url, method, headers, hashlib.sha256(data).hexdigest())
-    request = urllib.request.Request(url.geturl(), headers=headers, method=method, data=data)
-    try:
-        response = urllib.request.urlopen(request, context=host_ssl_context(url.netloc))
-    except urllib.error.HTTPError as exc:
-        logger.debug('%s %s %s → %s:', method, url.geturl(), headers, exc.status)
-        logger.debug('  %s', exc.read())
-        raise
+    tries = 3
+    while True:
+        headers = sign_request(url, method, headers, hashlib.sha256(data).hexdigest())
+        request = urllib.request.Request(url.geturl(), headers=headers, method=method, data=data)
+        try:
+            response = urllib.request.urlopen(request, context=host_ssl_context(url.netloc))
+        except urllib.error.HTTPError as exc:
+            logger.debug('%s %s %s → %s:', method, url.geturl(), headers, exc.status)
+            logger.debug('  %s', exc.read())
+            if exc.status == 503:
+                tries -= 1
+                if tries != 0:
+                    continue
```

Comment:
Should we really retry *immediately* here? Usually there's some exponential back-off, like `time.sleep(2**2**i)` (with `i = 3 - tries`, or you write the loop the other way around, etc.)

Reply:
I contemplated this.  I didn't do it, since this function isn't documented as "might sleep", but on the other hand, it's not exactly expected to be fast, either.

Reply:
I added a sleep of `10 ** (2 - tries)` which works out to 1s on the first iteration and 10s on the second.